include/swift/AST/LifetimeDependence.h

```diff
@@ -0,0 +1,110 @@
+//===--- LifetimeDependenceSpecifiers.h ------------------------*- C++ -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2024 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines types and utilities related to Lifetime Dependence
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef SWIFT_AST_LIFETIMEDEPENDENCE_H
+#define SWIFT_AST_LIFETIMEDEPENDENCE_H
+
+#include "swift/AST/Decl.h"
+#include "swift/Basic/Debug.h"
+#include "swift/Basic/OptionSet.h"
+#include "llvm/Support/TrailingObjects.h"
+
+namespace swift {
+
+enum class LifetimeDependenceKind : uint8_t {
+  Copy = 0,
+  Consume = 1,
+  Borrow = 2,
+  Mutate = 3
+};
+
+class LifetimeDependenceSpecifier {
+public:
+  enum class SpecifierKind { Named, Ordered, Self };
+
+private:
+  SourceLoc loc;
+  SpecifierKind specifierKind;
+  LifetimeDependenceKind lifetimeDependenceKind;
+  union Value {
+    struct {
+      Identifier name;
+    } Named;
+    struct {
+      unsigned index;
+    } Ordered;
+    struct {
+    } self;
+    Value(Identifier name) : Named({name}) {}
+    Value(unsigned index) : Ordered({index}) {}
+    Value() {}
+  } value;
```

Comment:
Instead of doing this could we take advantage of inline bit fields here? We'll get a NamedSpecifier, OrderedSpecifier or SelfSpecifier sublcasses and share trailing storage to store name/id similar to https://github.com/apple/swift/pull/70998?

Reply:
Can I do this as a follow on commit ?